pkg/morph/deploy/nns.go

```diff
@@ -226,3 +227,174 @@ func lookupNNSDomainRecord(inv *invoker.Invoker, nnsContract util.Uint160, domai
+			continue
+		}
+
+		// we could also try to compare on-chain version with the local one and tune
+		// update strategy: don't try to update with earlier version and catch updated
+		// contracts with unchanged version (blunder, we should react to it). For
+		// simplicity, we naively rely on the version change when the contract changes
+
+		extraUpdateArgs, err := prm.buildVersionedExtraUpdateArgs(versionOnChain)
+		if err != nil {
+			prm.logger.Error("failed to prepare build extra arguments for NNS contract update, will try again later",
+				zap.Stringer("on-chain version", versionOnChain), zap.Error(err))
+			continue
+		}
+
+		setGroupInManifest(&prm.localManifest, prm.localNEF, prm.committeeGroupKey, prm.localAcc.ScriptHash())
+
+		var vub uint32
+
+		// we pre-check 'already updated' case via MakeCall in order to not potentially
+		// wait for previously sent transaction to be expired (condition below) and
+		// immediately succeed
+		tx, err := committeeActor.MakeCall(nnsOnChainState.Hash, methodUpdate,
+			bLocalNEF, jLocalManifest, extraUpdateArgs)
+		if err != nil {
+			if isErrContractAlreadyUpdated(err) {
+				// FIXME: we get here when on-chain version is >= than the local one. If equals,
+				// this case must be considered as failure since version must be changed if code
+				// changes (NEF checksum differs here according to condition above). If greater,
+				// we should try to update at all. For simplicity, such condition is considered
+				// as success by current procedure now. To catch it, we need to pre-read version
+				// of the local contract using local NEF.
+				prm.logger.Info("NNS contract has already been updated, skip")
+				return nil
+			}
+		} else {
+			if updateTxValidUntilBlock > 0 {
+				prm.logger.Info("transaction updating NNS contract was sent earlier, checking relevance...")
+
+				if cur := prm.monitor.currentHeight(); cur <= updateTxValidUntilBlock {
+					prm.logger.Info("previously sent transaction updating NNS contract may still be relevant, will wait for the outcome",
+						zap.Uint32("current height", cur), zap.Uint32("retry after height", updateTxValidUntilBlock))
+					continue
+				}
+
+				prm.logger.Info("previously sent transaction updating NNS contract expired without side-effect")
+			}
+
+			prm.logger.Info("sending new transaction updating NNS contract...")
+
+			_, _, vub, err = committeeActor.Notarize(tx, nil)
+		}
+		if err != nil {
+			lackOfGAS := isErrNotEnoughGAS(err)
+			// here lackOfGAS can become true even if error corresponds to insufficient
+			// funds for the main transaction, not Notary balance problem. Right now we
+			// cannot distinguish between these two cases. So, we naively believe that the
+			// problem is in the Notary balance, because according to used procedure we make
+			// Notary deposit when at least 50GAS is available (most likely enough for all
+			// transactions).
```

Comment:
This statement is not true. Notarizer sends notary request to the notary pool, the pool cares only about sender's notary deposit amount (see [the link](https://github.com/nspcc-dev/neo-go/blob/abb58fb7de6053dc9734b5acfa334870481b6110/pkg/network/server.go#L1190)), thus the "insufficient funds" error can be returned only when sender doesn't have enough funds deposited to Notary contract.

And I saw your conversation with @roman-khimov in the NeoGo element chat, there probably was a misunderstanding.

Reply:
so, let me double-check, if `Notarize` returns `insufficient funds`, then its 100% Notary deposit problem?

Reply:
Exactly. You can get `insufficient funds` problem for main transaction only after it's completed by Notary service and Notary service is trying to add main transaction into standard mempool and relay. And even in this case you'll only see the following log message: `new transaction callback finished with error` warning `insufficient funds` error field.

Reply:
then i will split `MakeCall` and `Notatize` calls to distinguish GAS and Notary balance problems

Reply:
actually even simple actor returns `insufficient funds` only when we `Send`, not `Make` transaction

Reply:
fixed